### PR TITLE
feat: Cross-validation reporter with accessors

### PR DESCRIPTION
**Pull request by glemaitre**
I give some thoughts to https://github.com/probabl-ai/skore/issues/731.

I wanted to see if I could come with an API that I would be happy with.

So here, this is a class implementing aggressive caching to avoid recomputation in line with the idea of the current `cross_validate`. Actually, it can go further because, we could make it work with custom scorer/metric where we would not require to recompute the model predictions if already in the cache.

In terms of API, I went on the pandas side with accessors. I find it quite need to extend upon the current scikit-learn display or metrics. In terms of feature, none of the current scikit-learn displays offer visualization for cross-validation. I implemented a dirty trick to show that we could go quite fast from today without the need to wait too much on the scikit-learn side. Once ready in scikit-learn, we could of course leverage the proposed design there.

Also, I used the `available_if` trick to expose methods that are only relevant with the estimators and target provided. If trying to access something not compatible, it will raise and explained why this is not supported.

So at this stage, there is no test and far to be the cleaner code ever but it allows to have discussion in terms of design and additional features.

### CODE REVIEW

**Comment by adrinjalali** [comment]
yeah this looks more like it.

**Comment by glemaitre** [comment]
So some of those might be on the "cute" side but they are stuff that make me happy. I used `rich` for the help function:

image

This would restrain the output of what is available for the current cross-validation/estimator/data. Depending of the model, you will not get the same methods.

Instead of reusing, the `neg_` convention that is really difficult to explain to new comers, we could instead have the normal name of metric and use visual clues to mention if we deal with a loss or a score:

image

**Comment by adrinjalali** [comment]
You probably don't wanna use green and red though 😁

**Comment by glemaitre** [comment]
> You probably don't wanna use green and red though

Indeed :). Orange and blue are you friend

**Comment by MarieS-WiMLDS** [comment]
I like that it splits back and front. It means that if the user wants to play with CrossValidationReporter, they can do it immediately, they don't have to get an object with un unclear type, since they are not the ones who put it.

**Comment by glemaitre** [comment]
> and they'll have to pass clf again as well, to enable the use-case of tracking the model performance)

I'm not sure what this refers to here. Do you mean that `clf` would be fit and the full `X` and `y` (or single split)? This is seems unrelated to cross-validation in this case. If this is about cross-validation, I think that the `cv_results` will contain all the necessary estimators then.

**Comment by glemaitre** [comment]
> Also, with the second option, put is delayed until the last step, so until the user puts, they're in danger of losing their work; this is especially dangerous in a notebook.

Actually, I might be OK having both option:

```python
reporter = CrossValidationReporter(cv_results, X, y)
project.put("my_cv_results", reporter)
```

and

```python
reporter = CrossValidationReporter(cv_results, X, y, store=project)
```

Since the `CrossValidationReporter` is not anymore close to a scikit-learn function/method. I would still be ambivalent and let the user decide the way he wants it. I could imagine that the first way of doing, we could store everything in RAM until the `put` call while the second could allow to persist in the database/disk (no idea how skore is performing on this right now) directly.

**Comment by MarieS-WiMLDS** [comment]
## UX changes

Suggestion to avoid asking the user to do "return_xxx", asking them to give X and y again, and to allow for adding scorings as the current feature does.

With this piece of code, suggested by Guillaume in a discussion this afternoon: 

```python
class CrossValidationReporter:
  def __init__(self, cv_results, X, y):
    """You should probably use `from_estimator` or `from_cv_results`."""
    self.cv_results = cv_results
    self.X = X
    self.y = y
```

We would have this change.   

| before | after |
| --- | --- |
| <pre> from sklearn.model_selection import cross_validate <br> cv_results = cross_validate(classifier, X, y) </pre> | <pre> from skore import CrossValidationReporter <br> reporter = CrossValidationReporter(classifier, X, y) <br> cv_results = reporter.cv_results_ |

And the second option offers all the functions listed in the top comment.

This would answer the need: 
- change quite simple for the user, although not transparent
- possible for them to have exactly the same thing as before

## Suggestion about the "put" and parameter "project".

I'm in favor of having both. Setting `project = project` wouldn't put actually, but it would make it appear in the activity feed. This way, we avoid overloading the items list. So far, our testers are a bit surprised to see something appear in the items sections while they didn't put it, while they just learned that they can make items appear by doing using the put function. 

## Current features

Some of the current features are not included in this rework and will need to be added back: 
- add scorings (value: avoid having to rerun everything when you notice that you didn't chose correctly the scoring)
- aggregation: I think it cannot be included in the object CrossValidationReporter, it doesn't make sense. I think it would be more something like a new object, CrossValidationComparator.

## Questions remaining

imo  
- [ ] how does the object reporter appears in a view? What is displayed?
- [ ] what does the CrossValidationComparator? What are its inputs?

**Comment by sylvaincom** [comment]
Quote from @MarieS-WiMLDS:
> Two backends
> For now, following a decision taken this morning with the design team (cf this [notion page](https://www.notion.so/probabl/2024-11-22-skore-workshop-with-Ga-l-and-Guillaume-146ef76d36b9808d86cec6da33699d24) for probablers), we will keep only one backend, plotly, to keep it simple and easy to develop.

In his code, @glemaitre does `from sklearn.metrics import RocCurveDisplay` so the matplotlib plots are already coded inside of scikit-learn, we only need to add the plotly version ; so if we are to go with plotly, we might as well keep matplotlib (quick-win because in scikit-learn) and thus have both?

**Comment by thomass-dev** [comment]
@glemaitre your PR was automatically closed because we rebase the main branch in order to clean the documentation commits.

In addition, your issue is superseded by the issue #807. It will be the base of our implementation.
